examples/librilight/matching.py

```diff
@@ -0,0 +1,566 @@
+                        name=book_path,
+                        s=book_text,
+                        use_utf8=params.use_utf8,
+                    )
+                    books.append(book)
+
+            if not transcripts:
+                continue
+
+            logging.info(f"Thread[{thread_index}] loading cuts and books done.")
+
+            sourced_transcript_lists = texts_to_sourced_texts(
+                transcripts, uppercase=params.use_uppercase
+            )
+            sourced_transcripts = append_texts(sourced_transcript_lists)
+
+            sourced_book_list = texts_to_sourced_texts(
+                books, uppercase=params.use_uppercase
+            )
+            sourced_books = append_texts(sourced_book_list)
+
+            def _is_not_punc(c: np.int32) -> bool:
+                return not is_punctuation(chr(int(c)))
+
+            # Removing the punctuation
+            sourced_books = filter_texts(sourced_books, fn=_is_not_punc)
+
+            sourced_text = append_texts([sourced_transcripts, sourced_books])
+
+            logging.info(f"Thread[{thread_index}] construct sourced_text done.")
+
+            assert query_len == sourced_text.doc_splits[len(transcripts)], (
+                query_len,
+                sourced_text.doc_splits[len(transcripts)],
+            )
+            data_queue.put(
+                {
+                    "query_len": query_len,
+                    "cuts": batch_cuts,
+                    "cut_indexes": transcripts_cut_index,
+                    "sourced_text": sourced_text,
+                }
+            )
+        except Exception as e:
+            logging.error(f"Thread[{thread_index}] caught {type(e)}: e")
+            continue
+    barrier.wait()
+    if thread_index == 0:
+        data_queue.put(None)
+    logging.info(f"Thread[{thread_index}] dataloader done.")
+
+
+def aligner(
+    thread_index: int,
+    params: AttributeDict,
+    data_queue: Queue,
+    align_queue: Queue,
+    barrier: Barrier,
+    thread_pool: ThreadPool,
+):
+    """
+    Align each query in the sourced_text to its corresponding segment in references
+
+    The item in data_queue is:
+      {
+          "query_len": query_len,      # the total number of tokens of queries
+          "cuts": batch_cuts,          # A list of MonoCut from cuts_queue
+          "cut_indexes": transcripts_cut_index, # A tuple of (cut index, supervision index)
+          "sourced_text": sourced_text, # The sourced_text constructed from batch_cuts.
+      }
+
+    The item in align_queue is:
+      {
+          "cuts": item["cuts"],  # A list of MonoCut inherited from data_queue.
+          "cut_indexes": item["cut_indexes"], # Inherit from data_queue.
+          "alignments": alignments,  # The alignment results.
+          "sourced_text": sourced_text, # Inherit from data_queue.
+      }
+
+    """
+    while True:
+        try:
+            item = data_queue.get()
+            if item is None:
```

Comment:
i assume this is a kind of terminator, and we re-put for other threads to see it; can we comment this?